test/suite/Annotations.test.php

```diff
@@ -585,12 +592,72 @@ protected function testGetAnnotationsFromTrait()
             return;
         }
 
-        eval('trait TestTrait { }');
-        $this->setExpectedException(
-            self::ANNOTATION_EXCEPTION,
-            "Reading annotations from interface/trait 'TestTrait' is not supported"
-        );
-        Annotations::ofClass('TestTrait', '@note');
+        $annotations = Annotations::ofClass('SimpleTrait', '@note');
+
+        $this->check(count($annotations) === 1, 'SimpleTrait has one note annotation.');
+    }
+
+    protected function testCanGetMethodAnnotationsIncludedFromTrait()
+    {
+        if (version_compare(PHP_VERSION, '5.4.0', '<')) {
+            $this->pass();
+            return;
+        }
+
+        $annotations = Annotations::ofMethod(new \ReflectionClass('SimpleTraitTester'), 'runFromTrait');
+        $this->check(count($annotations) > 0, 'from class reflection and method name');
+
+        $annotations = Annotations::ofMethod(new \ReflectionMethod('SimpleTraitTester', 'runFromTrait'));
+        $this->check(count($annotations) > 0, 'from method reflection');
+
+        $annotations = Annotations::ofMethod(new SimpleTraitTester(), 'runFromTrait');
+        $this->check(count($annotations) > 0, 'from class object and method name');
+
+        $annotations = Annotations::ofMethod('SimpleTraitTester', 'runFromTrait');
+        $this->check(count($annotations) > 0, 'from class name and method name');
+    }
+
+    protected function testHandlesMethodAliasingWithTraits()
+    {
+        if (version_compare(PHP_VERSION, '5.4.0', '<')) {
+            $this->pass();
+            return;
+        }
+
+        $annotations = Annotations::ofMethod('mindplay\test\traits\NamespaceC\TraitTester', 'run');
+        $this->check($annotations[0]->note === 'trait-test-base', 'parent-inherited annotation goes first');
+        $this->check($annotations[1]->note === 'second-test-trait', 'trait-inherited annotation goes second');
+        $this->check($annotations[2]->note === 'trait-tester', 'non-inherited annotation goes third');
+
+        $annotations = Annotations::ofMethod('mindplay\test\traits\NamespaceC\TraitTester', 'conflict');
+        $this->check($annotations[0]->note === 'first-test-trait', 'correct trait is used when insteadof specified');
+
+        $annotations = Annotations::ofMethod('mindplay\test\traits\NamespaceC\TraitTester', 'conflictResolved');
+        $this->check($annotations[0]->note === 'second-test-trait', 'correct trait is used when aliased with as');
+
+        $annotations = Annotations::ofMethod('mindplay\test\traits\NamespaceC\TraitTester', 'runAgain');
+        $this->check($annotations[0]->note === 'trait-test-base', 'aliased methods still inherit from parent');
+        $this->check($annotations[1]->note === 'simple-trait', 'correct trait is used when aliased with as');
+    }
+
+    protected function testCanGetPropertyAnnotationsIncludedFromTrait()
```

Comment:
Test doesn't cover scenario, where class already has such property as in used trait.

I don't know what PHP does, but I see 2 scenarios currently:

1. if trait property is ignored, then do nothing special (annotation of class property + all parent class inherited annotations of this property are used)
2. if trait property wins and class property is ignored, then only take annotations from trait property
3. another case would be if 2 traits are declaring same property. Need to be checked which property is actually used by PHP and use it's annotations